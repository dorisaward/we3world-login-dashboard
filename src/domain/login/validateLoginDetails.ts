/**
 * Validates the login details submitted by the user.
 *
 * @param email The email address submitted by the user
 * @param password The password submitted by the user
 * @returns The error message to display to the user, or true on successful login
 */
export const validateLoginDetails = (
  email?: string,
  password?: string,
): string | true => {
  if (!email) {
    return 'Please enter your email address';
  }
  if (!emailRegex.test(email)) {
    return 'Please enter a valid email address';
  }
  if (!password) {
    return 'Please enter your password';
  }
  if (password.length < 8) {
    return 'Your password should be at least 8 characters long';
  }
  return true;
};

const emailRegex = new RegExp(/^[^\s@]+@[^\s@]+\.[^\s@]+$/);
