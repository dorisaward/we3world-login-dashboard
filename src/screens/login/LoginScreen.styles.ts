import {StyleSheet} from 'react-native';

export const styles = StyleSheet.create({
  container: {
    height: '100%',
    padding: 10,
    flex: 1,
  },
  header: {
    color: 'black',
    textAlign: 'center',
    marginBottom: 10,
  },
  textInput: {
    borderColor: 'black',
    borderWidth: 1,
    borderRadius: 50,
    marginBottom: 10,
    paddingHorizontal: 10,
  },
  errorMessage: {
    color: 'red',
    marginVertical: 10,
    fontWeight: 'bold',
  },
});
