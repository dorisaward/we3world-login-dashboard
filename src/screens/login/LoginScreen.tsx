import React, {useCallback, useState} from 'react';
import {Button, ScrollView, Text, TextInput, View} from 'react-native';
import {styles} from './LoginScreen.styles';
import {validateLoginDetails} from '../../domain/login/validateLoginDetails';

type Props = {
  /**
   * Unrenders the login screen and renders the {@link DashboardScreen}
   */
  navigateToDashboard: () => void;
};

export const LoginScreen = ({
  navigateToDashboard,
}: Props): React.JSX.Element => {
  const [email, setEmail] = useState<string | undefined>(undefined);
  const [password, setPassword] = useState<string | undefined>(undefined);
  const [errorMessage, setErrorMessage] = useState<string | undefined>(
    undefined,
  );

  const handleSubmitLoginDetails = useCallback(() => {
    const validation = validateLoginDetails(email, password);
    if (typeof validation === 'string') {
      setErrorMessage(validation);
    } else if (validation) {
      setErrorMessage(undefined);
      navigateToDashboard();
    }
  }, [email, password, setErrorMessage, navigateToDashboard]);

  return (
    <ScrollView contentInsetAdjustmentBehavior="automatic">
      <View style={styles.container}>
        <Text style={styles.header}>Please enter your email and password</Text>
        <TextInput
          style={styles.textInput}
          value={email}
          placeholder="email address"
          onChangeText={setEmail}
        />
        <TextInput
          style={styles.textInput}
          secureTextEntry={true}
          value={password}
          placeholder="password"
          onChangeText={setPassword}
        />
        {errorMessage && (
          <Text style={styles.errorMessage}>{errorMessage}</Text>
        )}
        <Button title="login" onPress={handleSubmitLoginDetails} />
      </View>
    </ScrollView>
  );
};
