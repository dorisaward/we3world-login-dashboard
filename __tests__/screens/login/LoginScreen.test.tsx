import React from 'react';
import {it, expect} from '@jest/globals';
import {LoginScreen} from '../../../src/screens/login/LoginScreen';
import {fireEvent, render} from '@testing-library/react-native';
import {validateLoginDetails} from '../../../src/domain/login/validateLoginDetails';

jest.mock('../../../src/domain/login/validateLoginDetails');

it('should render', () => {
  // Given
  const renderable = <LoginScreen navigateToDashboard={jest.fn()} />;

  // When
  const {toJSON} = render(renderable);

  // Then
  expect(toJSON()).toMatchSnapshot();
});

it('should display error message, given login details not valid', () => {
  // Given
  const errorMessage = 'test error message';
  jest.mocked(validateLoginDetails).mockImplementationOnce(() => errorMessage);

  const navigateToDashboard = jest.fn();
  const renderable = <LoginScreen navigateToDashboard={navigateToDashboard} />;
  const {getByText, queryByText} = render(renderable);
  const loginButton = getByText('login');

  // When
  fireEvent.press(loginButton);

  // Then
  expect(navigateToDashboard).not.toHaveBeenCalled();
  expect(queryByText(errorMessage)).toBeTruthy();
});

it('should navigate, given login details valid', () => {
  // Given
  jest.mocked(validateLoginDetails).mockImplementationOnce(() => true);

  const navigateToDashboard = jest.fn();
  const renderable = <LoginScreen navigateToDashboard={navigateToDashboard} />;
  const {getByText} = render(renderable);
  const loginButton = getByText('login');

  // When
  fireEvent.press(loginButton);

  // Then
  expect(navigateToDashboard).toHaveBeenCalled();
});
