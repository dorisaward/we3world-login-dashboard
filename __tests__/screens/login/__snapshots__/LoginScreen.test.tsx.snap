// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render 1`] = `
<View
  style={
    {
      "flex": 1,
      "height": "100%",
      "padding": 10,
    }
  }
>
  <Text
    style={
      {
        "color": "black",
        "marginBottom": 10,
        "textAlign": "center",
      }
    }
  >
    Please enter your email and password
  </Text>
  <TextInput
    onChangeText={[Function]}
    placeholder="email address"
    style={
      {
        "borderColor": "black",
        "borderRadius": 50,
        "borderWidth": 1,
        "marginBottom": 10,
        "paddingHorizontal": 10,
      }
    }
  />
  <TextInput
    onChangeText={[Function]}
    placeholder="password"
    secureTextEntry={true}
    style={
      {
        "borderColor": "black",
        "borderRadius": 50,
        "borderWidth": 1,
        "marginBottom": 10,
        "paddingHorizontal": 10,
      }
    }
  />
  <View
    accessibilityRole="button"
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        [
          {},
        ]
      }
    >
      <Text
        style={
          [
            {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        login
      </Text>
    </View>
  </View>
</View>
`;
